packages/select/src/components/query-list/queryList.tsx

```diff
@@ -267,14 +265,45 @@ export class QueryList<T> extends React.Component<IQueryListProps<T>, IQueryList
         Utils.safeInvoke(onKeyUp, event);
     };
 
-    private moveActiveIndex(direction: number) {
-        // indicate that the active item may need to be scrolled into view after update.
-        // this is not possible with mouse hover cuz you can't hover on something off screen.
-        this.shouldCheckActiveItemInViewport = true;
+    /**
+     * Get the next enabled item, moving in the given direction from the current
+     * index. An `undefined` return value means no suitable item was found. The
+     * second and third parameters are used internally for recursion.
+     * @param direction amount to move in each iteration, typically +/-1
+     * @param currentIndex current index being considered
+     * @param startIndex index at which recursion began, used to detect infinite loop
+     */
+    private getNextActiveItem(
+        direction: number,
+        currentIndex = this.getActiveIndex(),
+        startIndex = currentIndex,
+    ): T | undefined {
         const { filteredItems } = this.state;
-        const maxIndex = Math.max(filteredItems.length - 1, 0);
-        const nextActiveIndex = Utils.clamp(this.getActiveIndex() + direction, 0, maxIndex);
-        Utils.safeInvoke(this.props.onActiveItemChange, filteredItems[nextActiveIndex]);
+        if (filteredItems.length < 2) {
+            return filteredItems[currentIndex];
+        }
+
+        const nextActiveIndex = wrapNumber(currentIndex + direction, 0, filteredItems.length - 1);
+        if (nextActiveIndex === startIndex) {
+            // loop detected! we've returned to the start without finding a suitable candidate.
+            return undefined;
+        } else if (this.isItemDisabled(filteredItems[nextActiveIndex], nextActiveIndex)) {
+            // keep on moving in given direction if this item is disabled.
+            return this.getNextActiveItem(direction, nextActiveIndex, startIndex);
```

Comment:
seems a little weird to use recursion on an unbounded list when iterating would suffice

Reply:
i found recursion to be the simplest code change here (tho i then went ham on the refactor 🤷‍♂️)

Reply:
will refactor to use a loop